Use Class Names for Object Detection

ongtw opened the pull request:
Upgraded `DeclarativeLoader` to support using labels in addition to object IDs for object detection. Closes #488.
 
Instead of `detect_ids: [0, 1, 2, 3]`, can declare as `detect_ids: [person, bicycle, car, motorcycle]`.
Enclosing labels in quotes is entirely optional, like `detect_ids: ["person", "bicycle", "car", "motorcycle"]`.
The labels follow the `Class name` in the [Object Detection Models](https://peekingduck.readthedocs.io/en/stable/resources/01a_object_detection.html) page strictly, i.e. `people`, `cars` and `bike` are not recognised. Any duplicate/erroneous/unknown labels are ignored.
The `test_declarativeloader.py` file is also updated with the appropriate unit tests.

Technically, the main changes are:
- Add 2 new files `peekingduck/utils/mapping_yolo.txt` and `peekingduck/utils/mapping_efficientdet.txt` to map labels to the corresponding object IDs.
- In `DeclarativeLoader._edit_config()`, check if node is either Yolo or EfficientDet and key is `detect_ids`: if yes, call `DeclarativeLoader._change_labels_to_ids()`.
- `DeclarativeLoader._change_labels_to_ids()` calls `DeclarativeLoader._load_mapping()` to load the appropriate mapping text file and maps any labels to their corresponding object IDs.
- In short, all the (pre-)processing takes place at the loader portion and nothing is changed beneath the hood. The model will see numeric object IDs as per normal.

Undocumented feature: mixing labels and object IDs are allowed, e.g. `detect_ids: [0, car, 3, cell phone, laptop]` is legit, but not encouraged.

**Tech Notes**
- From [Coco paper](https://arxiv.org/abs/1405.0312): `Note that we have limited the 2014 release to a subset of 80 categories. We did not collect segmentations for the following 11 categories: hat, shoe, eyeglasses (too many instances), mirror, window, door, street sign (ambiguous and difficult to label), plate, desk (due to confusion with bowl and dining table, respectively) and blender, hair brush (too few instances). We may add segmentations for some of these categories in the cumulative 2015 release.`
- COCO paper and EfficientDet use `airplane`, but Yolo uses `aeroplane`.

leeping-ng commented on peekingduck/utils/mapping_efficientdet.txt:
```diff
@@ -0,0 +1,82 @@
+    "teddy bear": 87,
+    "hair drier": 88,
+    "toothbrush": 89,
+}
```
Would it be better for these mapping files to be in their respective model directories instead of `utils`? For consideration when @liyier90 reviews this PR.

ongtw replied:
TBH, I'm not 100% sure yet which is a good location for these files, based on the following considerations:

1. Store mapping files with models
- Cleaner organisation (+)
- Same (duplicate) mappings for multiple models, e.g. Yolov4 and YoloX (-)

2. Store mapping files in `utils` (or somewhere else)
- One file for multiple models, no duplication (c.f. above) (+)
- Cleaner implementation code for reading mapping files (+?)
- File organisation not as clean (-)

Welcome anyone to chip in to this discussion.

liyier90 replied:
I'm leaning towards not keeping them with the individual models to avoid having duplicated files.

To mirror my comment below, I'm also wondering if it's better to categorize the mapping as maybe "coco original numbering" (efficientdet) vs "coco condensed" (yolo, yolox). "coco condensed" is basically when they "shift the IDs up" so there's no skipped numbers.

liyier90 replied:
Refactored this into a single master mapping file.

liyier90 commented on peekingduck/declarative_loader.py:
```diff
@@ -180,12 +178,69 @@ def _edit_config(
         return dict_orig
 
+    def _change_labels_to_ids(
+        self, node_name: str, key: str, value: list
+    ) -> Tuple[str, list]:
```
List types should be `List[<type>]`

ongtw replied:
Corrected.

liyier90 commented on peekingduck/declarative_loader.py:
```diff
@@ -180,12 +178,69 @@ def _edit_config(
+        Returns:
+            Tuple[str, list]: "detect_ids", list of sorted ids for object detection
+        """
+        assert key == "detect_ids", f"Key Error: expect detect_ids but got {key}"
```
Is this necessary considering it only gets called during `if key == "detect_ids"`?

ongtw replied:
Removed `assert` line to reduce code. Wanted to "future-proof" `_change_labels_to_ids` but given the restricted scope in which this method is called, it's ok to not do extra checks (for now).

liyier90 commented on peekingduck/declarative_loader.py:
```diff
@@ -180,12 +178,69 @@ def _edit_config(
+            node_name (str): to determine if node is yolo or efficientdet as both
+                             lists of ids are different.
+            key (str): can only be "detect_ids", else error.
+            value (list): list of labels/ids for detection.
```
If we are certain enough that this function only handles converting labels to IDs, perhaps can make the variable names more explicit? Maybe something like `labels_and_ids`

ongtw replied:
`_change_labels_to_ids` is a function that maps:
1. string to int (1-to-1)
2. int to int (ID)

As the latter is essentially an identity function, I think it's ok to name it as such. It is verbose to call it `_change_labels_and_ids_to_ids`.

liyier90 replied:
I was referring the `value` argument/variable, thinking that naming it to `labels_and_ids` might make `for x in labels_and_ids` in the list comprehension more descriptive. But i guess the current naming is fine since that line of code is pretty close to the documentation

ongtw replied:
Oic. The name `value` is meant to reflect the `key`/`value` pair of the node attribute (specifically, `key=detect_ids`, `value=list of ids/labels`), and these two are returned to caller `DeclarativeLoader` to replace the current `key`/`value` pair being processed under [`_edit_config`](https://github.com/ongtw/PeekingDuck/blob/465312febe0b4ca4b9a898225a238690ab414647/peekingduck/declarative_loader.py#L185).

liyier90 commented on peekingduck/declarative_loader.py:
```diff
@@ -180,12 +178,69 @@ def _edit_config(
+        map_labels_ids = self._load_mapping(node_name)
+        # Use set to eliminate duplicates and/or errors (which are mapped to zero)
+        obj_ids_set = {
+            x if isinstance(x, int) else map_labels_ids[x] if x in map_labels_ids else 0
```
I think the 2nd part can be more concisely written as 
```suggestion
            x if isinstance(x, int) else map_labels_ids.get(x, 0)
```
relevant [SO](https://stackoverflow.com/questions/11041405/why-dict-getkey-instead-of-dictkey)

ongtw replied:
Thanks, `get(x, [default])` is nice! Corrected the code.

liyier90 commented on peekingduck/declarative_loader.py:
```diff
@@ -180,12 +178,69 @@ def _edit_config(
+        mapping_file = self.pkd_base_dir / "utils" / f"{filename}.txt"
+        with open(mapping_file, "r") as file:
+            contents = file.read()
+        the_mapping = ast.literal_eval(contents)
```
any particular advantage of using `ast.literal_eval` instead of writing the mapping in yaml and using `yaml.safe_load`?

Just trying to see if it's possible to only use one type of config/input related file type.

ongtw replied:
Good point. I have changed the filetype to `yml` and updated the code to read `yaml` file instead.

liyier90 commented on peekingduck/declarative_loader.py:
```diff
@@ -180,12 +178,69 @@ def _edit_config(
+            Dict[str, int]: mapping of labels to object ids relevant to given model
+        """
+        assert node_name in (
+            "model.yolo",
+            "model.efficientdet",
```
Similarly, is this necessary since a previous check has been done.

If it's necessary, to reduce maintenance burden, is it possible to keep a list of `model.<model_name>` in some txt or yaml file and read from that instead?

ongtw replied:
I think this check is necessary, to ensure that if future models also have a `detect_ids` key, that this code is properly updated for them.

liyier90 reviewed (changes requested):
Some additional considerations.

Does `_change_labels_to_ids` and `_load_mapping` have to be methods of `DeclarativeLoader`? Can it be placed in a separate script in maybe `utils` and have `pkd_base_dir` passed in as an argument if 

Some considerations regarding maintenance:
1. Will it be better/easier to define a single mapping (COCO classes) file with multiple mapping values for each class? For example,

```yaml
"stop sign":
  efficientdet: 12,
  yolo: 11
```

2. Will it be better to try and categorize these 2 ways of mapping into something more general than model names? The current way seems to suggest that we will need as many mapping files as there are COCO based object detectors.

Maybe something like `coco` and `coco_condensed`? `coco` for mapping to the [original IDs](https://tech.amikelive.com/node-718/what-object-categories-labels-are-in-coco-dataset/) and `coco_condensed` for the mapping that doesn't skip any numbers.

ongtw commented:
I like Yier's idea of having a single (master) mapping file. Maybe we can discuss this after the standup?

ongtw commented:
Recap as discussed during meeting: to implement a master mapping file in `peekingduck/pipeline/nodes/models` folder.
1. This ensures the mapping file is visible to developers and serves as a reminder to update it for future object detection models.
2. To refactor this in future if the need arises.
3. This mapping file could be extended in future to support other types of models, e.g. keypoint mappings (my post-meeting input).

ongtw commented:
Implemented the master mapping file as discussed by
1. Putting two yaml documents into one `peekingduck/pipeline/nodes/model/master_map.yml`: a) doc 1 tells which mapping system to use for which model, b) doc 2 stores the mapping from class labels to object IDs according to the different systems.
2. Shifted the relevant methods from DeclarativeLoader to ConfigLoader, where they make more sense (since they deal with model configuration).
3. Added relevant unit tests for ConfigLoader.

p.s. both `aeroplane` and `airplane` are admissible now.

liyier90 commented on peekingduck/configloader.py:
```diff
@@ -43,6 +46,46 @@ def _get_config_path(self, node: str) -> Path:
+            1. model mapping: tells which mapping system a particular object detection model uses
+            2. class name to ID mapping: maps class name to ID, supports multiple mapping systems
+        """
+        assert node_name in (
+            "model.yolo",
+            "model.efficientdet",
+        ), f"Name Error: expect model.yolo or model.efficientdet but got {node_name}"
```
Currently this only allows `yolo` and `efficientdet` but the master mapping yaml suggests that it works with `yolox` as well. Would it be better to check against the keys in `model_map_dict` instead of hard coding it here?

ongtw replied:
Good idea: generated `node_name_list` from `model_map_dict` and checked against that instead of hardcoding. Easier on maintenance too.

ongtw replied:
Haven't tested with `model.yolox` yet; am assuming it will work since it uses same numbering system as `yolo`.

liyier90 commented on tests/loaders/test_configloader.py:
```diff
@@ -62,3 +62,95 @@ def test_config_loader_load_correct_yaml(self, configloader):
+        test_res = configloader.change_class_name_to_id(node_name, key, val)
+        assert test_res == ground_truth
+
+    def test_config_loader_change_class_name_to_id_yolo_mix_int_text_errors(
```
After seeing this test case, I realize that we are silently defaulting to `detect_id=0` when a parse/conversion error occurs. Would it be better to log some kind of warning to the user, maybe something like

Could not convert class name "`<some name>`", defaulting to "person"...

Can help the user in troubleshooting in the cases of spelling errors or capitalizing the name by mistake

ongtw replied:
Yeah, the code defaults any unknown/erroneous class name to `0` since `0` is the default `detect_ids` value.
Updated code to lowercase all class names to avoid capitalization errors.

Can print out a list of erroneous class names to assist in user debugging.

ongtw replied:
Have updated the codebase to log print a list of invalid class names.

ongtw commented:
Codebase updated with latest comments above, except for testing with YoloX.

liyier90 commented on tests/loaders/test_configloader.py:
```diff
@@ -62,3 +62,95 @@ def test_config_loader_load_correct_yaml(self, configloader):
+        self, configloader
+    ):
+        node_name = "model.yolo"
+        test_map = configloader._load_mapping(node_name)
```
`test_map` for all test functions other than `test_config_loader_load_mapping_efficientdet` and `test_config_loader_load_mapping_yolo` are unused, consider removing

ongtw replied:
Remove unused `test_map` variables.

liyier90 commented on tests/loaders/test_configloader.py:
```diff
@@ -62,3 +62,95 @@ def test_config_loader_load_correct_yaml(self, configloader):
+        assert test_map["parking meter"] == 13
+        assert test_map["toothbrush"] == 89
+
+    def test_config_loader_load_mapping_yolo(self, configloader):
+        node_name = "model.yolo"
```
Since this is going into node-v1.2 which already contains yolox, consider adding tests for mapping for yolox model. Perhaps a fixture like
```
@pytest.fixture(params=["model.yolo", "model.yolox"])
def yolo_node_name(request):
    return request.param
```
So we can just increment that for more models which uses the yolo numbering in the future instead having to add new test functions for each new model.

Applicable to the other yolo related mapping tests below.

ongtw replied:
My working branch is not sync with node-v1.2, so doesn't have YoloX. Merging with node-v1.2 will introduce many changes.
Suggest to postpone writing the YoloX unit tests until this has been merged into node-v1.2 (keep this as open issue).

liyier90 replied:
If you meant introducing changes that will show as "Files Changed" in this PR, merging the latest node-v1.2 (target branch) should not do that. But sure, I think we can write the YOLOX tests after this is merged.

liyier90 commented on peekingduck/declarative_loader.py:
```diff
@@ -180,6 +178,14 @@ def _edit_config(
                         f"Config for node {node_name} does not have the key: {key}"
                     )
                 else:
+                    if key == "detect_ids" and node_name in (
+                        "model.yolo",
+                        "model.efficientdet",
+                    ):
```
I briefly tested and found the mapping to be working after adding `"model.yolox"` here. To echo the comment in the test file, consider adding yolox here since this change is going into node-v1.2.

Also, perhaps it's possible to use something similar to `node_name_list` in `configloader`? Unless it is intentional to hardcode the models here to ensure we've developed/tested the feature.

ongtw replied:
Remove `node_name` verification here and outsource the responsibility to the callee function `self.config_loader.change_class_name_to_id` instead.
This assumes the master map file is updated after a new model has been developed and tested.

liyier90 reviewed (changes requested):
Just one final change to make this feature work with YOLOX since this PR is going into `node-v1.2`. Other than that it looks good to go.

liyier90 reviewed (approved):
Outstanding task (to be done after merge): Write unit tests for mapping IDs for YOLOX.
